[FEATURE] New TracingGanttChart panel

andreasgerstmayr opened the pull request:
# Description
This PR adds a new `TracingGanttChart` panel. This is built from scratch, based on React and Material UI.

Resolves #978
Obsoletes #2033

**Features**
* colors are generated deterministically (same as the `TimeSeriesChart`)
* spans with errors have a red hue and an error icon next to the name
* span collapsing
* span indention guide highlighting
* resizable divider (between span name and span bars)
* dynamic indention width based on the spans in viewport (making it easier to navigate deeply nested spans)
* ticks at 0%, 25%, 50%, 75% and 100%
* mini Gantt chart overview to select a time frame of the trace to zoom into
* sidepanel: span attributes and events are visible in a side panel (after clicking on a span)
* performance: tested with traces up to 800 spans; `react-virtuoso` only renders rows which are visible
* unit tests

**Alternative Gantt chart libraries considered**
* https://scheduler.bitnoise.pl/
* https://frappe.io/gantt
* https://github.com/MaTeMaTuK/gantt-task-react
* https://www.npmjs.com/package/vis-timeline
* https://anovokmet.github.io/svelte-gantt
* https://thetechcompany.github.io/react-timeline-gantt/
* https://github.com/robicch/jQueryGantt

Some of them look nice, but all of them are basically intended for project planning purposes (using days and weeks, drag 'n' drop support, etc.). Tweaking one of these to match our use case is likely more effort than starting from scratch.

# Screenshots

![grafik](https://github.com/perses/perses/assets/538011/acce4e17-dc1a-4813-8bcd-74ffe8111fae)

https://github.com/perses/perses/assets/538011/f7e2692c-1ee4-4fc0-b2b4-d612ba702a2e

![grafik](https://github.com/perses/perses/assets/538011/2682a4f0-ec55-4997-ad1d-2cca23cb59a5)

# Checklist

- [X] Pull request has a descriptive title and context useful to a reviewer.
- [X] Pull request title follows the `[<catalog_entry>] <commit message>` naming convention using one of the
  following `catalog_entry` values: `FEATURE`, `ENHANCEMENT`, `BUGFIX`, `BREAKINGCHANGE`, `DOC`,`IGNORE`.
- [X] All commits have [DCO signoffs](https://github.com/probot/dco#how-it-works).

## UI Changes

- [X] Changes that impact the UI include screenshots and/or screencasts of the relevant changes.
- [X] Code follows the [UI guidelines](https://github.com/perses/perses/blob/main/ui/ui-guidelines.md).
- [ ] Visual tests are stable and unlikely to be flaky.
  See [Storybook](https://github.com/perses/perses/tree/main/ui/storybook#visual-tests)
  and [e2e](https://github.com/perses/perses/tree/main/ui/e2e#visual-tests) docs for more details. Common issues
  include:
  - Is the data inconsistent? You need to mock API requests.
  - Does the time change? You need to use consistent time values or mock time utilities.
  - Does it have loading states? You need to wait for loading to complete.

andreasgerstmayr commented on ui/panels-plugin/src/plugins/trace-table/DataTable.tsx:
```diff
@@ -70,21 +73,34 @@ export function DataTable({ result }: DataTableProps) {
 
+  const traceLinkParams = new URLSearchParams({
+    explorer: '1',
+    queries: `[{"kind":"TraceQuery","spec":{"plugin":{"kind":"TempoTraceQuery","spec":{"query":"${trace.traceId}"}}}}]`,
```
This seems hacky, is there a better way to do it?
I could use `onClick`/`setQueries` and handle it client side, but then I'll loose the browser features of real links, like open a link in a new (background) tab, etc.

Gladorme replied:
Hmm, I don't think of a better solution :/

However, I think we should remove this magic number (1) and put it behind a variable 😄

andreasgerstmayr replied:
In which package should I define this constant? It belongs to the `explore` package, however we don't have a dependency from `panels-plugin` on the `explore` package (I'm not sure if I should add one? Doesn't seem right).

Maybe we can change `explorer=1` to `explorer=traces`, to avoid having a magic number in the code, and not need a dependency from `panels-plugin` to the `explore` package?

andreasgerstmayr commented on ui/explore/src/components/ExploreManager/TracesExplorer.tsx:
```diff
@@ -32,55 +90,21 @@ function TracePanel({ queries }: { queries: QueryDefinition[] }) {
-  const { tab, queries, setTab, setQueries } = useExplorerManagerContext();
+  // TODO: how to access query string here?
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  const isSingleTrace = definitions.length === 1 && isValidTraceId((definitions[0]?.spec as any).query || '');
```
How can I access the `query` here? It's required to determine whether to show the trace table or Gantt chart.

Gladorme replied:
How to access `query` without the `as any` ?

andreasgerstmayr replied:
I think it's not possible, because the required type `TempoTraceQuerySpec` is in `tempo-plugin`, which is not (and should not be) a dependency of `@perses-dev/explore`. I've extended the condition to check the `kind`s, but I guess that's all I can do here :)

... unless we would move the `query` one level up to the `TraceQuery` instead of `TempoTraceQuery`. But I guess it's ok as-is now.

AntoineThebaud reviewed (changes requested):
Could you please add the associated cuelang schema on the backend side, just like you did for TraceTable in https://github.com/perses/perses/pull/2061? Thanks in advance :)

andreasgerstmayr commented on ui/panels-plugin/src/plugins/trace-table/trace-table-model.ts:
```diff
@@ -11,16 +11,29 @@
 // eslint-disable-next-line @typescript-eslint/no-empty-interface
-export interface TraceTableOptions {}
+export interface TraceTableOptions {
+  onTraceClick?: (e: MouseEvent, traceId: string) => void;
```
Is this fine? I think in general the `spec` of a Panel should be JSON-serializable, but in this case I need functions for
* running `setQueries()` on the explore page
* creating a link to the trace detail page (useful when the panel gets embedded outside of perses)

Gladorme replied:
Maybe it's better to do it in `TraceTableProps` and keep `TraceTableOptions` interface to model stuff?
```ts
export interface TraceTableProps extends PanelProps<TraceTableOptions> {
  onTraceClick?: (e: MouseEvent, traceId: string) => void;
  traceLink?: (traceId: string) => string;
}
``` 

andreasgerstmayr replied:
I like this idea :+1:

If these properties should be set, this panel (React component) must be embedded from code anyway, and it's not possible to set this property from a JSON dashboard specification.

Gladorme commented:
Hey, sorry for the delay. Thanks for the local env setup 👍 
However, I tested / played with our PRD Tempo. It's really great! 😄 

I spotted few issues:
- Dark theme have some CSS color issue
  - On the trace time line: 
![image](https://github.com/user-attachments/assets/ca0c75a6-af5c-44d4-a875-2b87b030651a)
  - Spans char on span hover:
![image](https://github.com/user-attachments/assets/f73c4128-6349-40cb-ac5a-3caad3ba401e)
  - Traces table on trace hover: 
![image](https://github.com/user-attachments/assets/c584892d-ea60-4cc3-80fe-c5cfb2b372dd)
- Some data are truncated on the right and bottom of spans table:
![image](https://github.com/user-attachments/assets/e4f5de45-946c-447d-9b89-ebd5b0b4de07)
- When users do not chose the default datasource and click on a trace on trace table. It will redirect to the span panel with the default datasource

And some futur improvements:
- Trace table loading: once the "search" request is done maybe we could show the table, instead of waiting all traces requests. For example, the search request was finished at 1.5s, but the all traces requests took 9s to finish
![image](https://github.com/user-attachments/assets/ef1a1937-7296-4806-a319-289303c687b5)
- Do not show "Events" tab if there is no event and/or show a chip with the number of events in the span detail view
![image](https://github.com/user-attachments/assets/b5f3d7b9-4934-4e0e-98c8-a0b8a4c6c722)
- Allow users to resize span detail view

Gladorme commented on ui/explore/src/components/ExploreManager/TracesExplorer.tsx:
```diff
@@ -32,55 +105,28 @@ function TracePanel({ queries }: { queries: QueryDefinition[] }) {
-
-export function TracesExplorer() {
-  const { tab, queries, setTab, setQueries } = useExplorerManagerContext();
+  // Cannot cast to TempoTraceQuerySpec because 'tempo-plugin' types are not accessible in @perses-dev/explore
```
It will be easier when the explorer components will be located in plugin too.

Gladorme commented on ui/panels-plugin/src/plugins/trace-table/trace-table-model.ts:
```diff
@@ -11,16 +11,29 @@
  * The Options object type supported by the TraceTable panel plugin.
  */
 // Note: The interface attributes must match cue/schemas/panels/trace-table/trace-table.cue
 // eslint-disable-next-line @typescript-eslint/no-empty-interface
```
`// eslint-disable-next-line` can be remove if we keep it like that (depends on what is done for next comment 😄 )

andreasgerstmayr commented:
Thank you for the review!

> Dark theme have some CSS color issue

Ah, good catch. I forgot to test the dark theme.
Fixed.

>  Some data are truncated on the right and bottom of spans table:

Should be fixed now.
 
>  When users do not chose the default datasource and click on a trace on trace table. It will redirect to the span panel with the default datasource

Fixed.

> And some futur improvements:
> 
>  Trace table loading: once the "search" request is done maybe we could show the table, instead of waiting all traces requests. For example, the search request was finished at 1.5s, but the all traces requests took 9s to finish

This will be resolved once Tempo moves to vParquet 4. It's already the default block format upstream, but there's no release yet including this change.

>  Not show "Events" tab if there is no event and/or show a chip with the number of events in the span detail view

Done.

>  Allow users to resize span detail view

Done.

andreasgerstmayr commented:
There's one minor issue remaining: when you click on a link in the trace table, it does a full page refresh.
If I remove `reloadDocument` from the `<Link>`, it updates the URL but doesn't update the content due to how the state is propagated from `<ExplorerManagerProviderWithQueryParams>` to `<ExplorerManagerProvider>`. I'll create a separate PR for it.

Gladorme reviewed (approved):
LGTM

Gladorme commented:
Minor issue: if you go to event tab with a span that have an event and click on a span without event. There is nothing in the view and no tab selected. Maybe we should still display event tab if it's the current tab even if there is no event OR switch tab to attributes if there is no event 🤔

andreasgerstmayr commented:
> Minor issue: if you go to event tab with a span that have an event and click on a span without event. There is nothing in the view and no tab selected. Maybe we should still display event tab if it's the current tab even if there is no event OR switch tab to attributes if there is no event 🤔

Good catch! Fixed, I'm switching the tab state to attributes now if there's no event.
